src/DataAPI.jl

```diff
@@ -287,4 +287,100 @@ using a `sink` function to materialize the table.
 """
 function allcombinations end
 
+const STYLE_INFO = """
+One of the uses of the metadata style information is decision
+how the metadata should be propagated when `x` is transformed. This interface
+defines `:none` style that indicates that metadata should not be propagated
+under transformations. At least this style must be supported by any type
+defining support for metadata.
+"""
+
+"""
+    metadata(x, key, [default]; full::Bool=false)
+
+Return metadata value associated with object `x` for key `key`.
+If `x` does not support metadata throw `ArgumentError`.
+If `x` supports metadata, but does not have a mapping for `key` throw `KeyError`.
+
+If `full=true` return a tuple of metadata value and metadata style. Metadata
+style is an additional information about the kind of metadata that is stored
+for the `key`.
+
+$STYLE_INFO
+
+If `default` is passed then return it if `x` does not support metadata or
+does not have a mapping for `key`. The style of `default` value is `:none`.
+"""
+metadata(::T, ::Any; full::Bool=false) where {T} =
+    throw(ArgumentError("Objects of type $T do not support getting metadata"))
+metadata(::Any, ::Any, default; full::Bool=false) = full ? (default, :none) : default
+
+"""
+    metadatakeys(x)
+
+Return an iterator of metadata keys for which `metadata(x, key)` returns a
+metdata value. If `x` does not support metadata return `()`.
+"""
+metadatakeys(::Any) = ()
+
+"""
+    metadata!(x, key, value; style)
+
+Set metadata for object `x` for key `key` to have value `value` and style `style`
+and return `x`.
+If `x` does not support setting metadata throw `ArgumentError`.
+
+$STYLE_INFO
+"""
+metadata!(::T, ::Any, ::Any; style) where {T} =
+    throw(ArgumentError("Objects of type $T do not support setting metadata"))
+
+"""
+    colmetadata(x, col, key, [default]; full::Bool=false)
+
+Return metadata value associated with table `x` for column `col` and key `key`.
+If `x` does not support metadata for column `col` throw `ArgumentError`. If `x`
+supports metadata, but does not have a mapping for column `col` for `key` throw
+`KeyError`.
+
+If `full=true` return a tuple of metadata value and metadata style. Metadata
+style is an additional information about the kind of metadata that is stored for
+the `key`.
+
+$STYLE_INFO
+
+If `default` is passed then return it if `x` does not support metadata for
+column `col` or does not have a mapping for `key` for column `col`. The style of
+`default` value is `:none`.
+"""
+colmetadata(::T, ::Any, ::Any; full::Bool=false) where {T} =
+    throw(ArgumentError("Objects of type $T do not support getting column metadata"))
+colmetadata(::Any, ::Any, ::Any, default; full::Bool=false) = full ? (default, :none) : default
```

Comment:
A potential issue with this fallback is that it won't throw an error if the column doesn't exist. That means that when a type implements metadata, it won't be able to throw in that case either as users may rely on this behavior. I see only two possibilities:
- document this behavior in the docstring (i.e. require implementations to follow it) by adding something like "if `x` does not have column `col`"
- throw an error when type does not support metadata

Reply:
This is also something I was not sure what to do. I think, given your comment, I will indeed throw an error if the column does not exist.